File: unittest/body.cpp
/*
 * Copyright 2014, Nicolas Mansard, LAAS-CNRS
 *
 * This file is part of eigenpy.
 * eigenpy is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 * eigenpy is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.  You should
 * have received a copy of the GNU Lesser General Public License along
 * with eigenpy.  If not, see <http://www.gnu.org/licenses/>.
 */

#define EIGEN_DONT_VECTORIZE
#define EIGEN_DISABLE_UNALIGNED_ARRAY_ASSERT 

#include "rbdlpy/rbdlpy.hpp"
#include <boost/python.hpp>

using namespace boost::python;
using namespace std;
using namespace RigidBodyDynamics;
using namespace eigenpy;

// Curiously Recursive Template Patern (CRTP) for exposing Body class
struct Body_visitor : boost::python::def_visitor<Body_visitor>
{
    friend class def_visitor_access;

    template <class classT>
    void visit(classT& c) const
    {
        c   .add_property("mass", &Body_visitor::get_mMass, &Body_visitor::set_mMass)
            .add_property("centerOfMass", &Body_visitor::get_mCenterOfMass, &Body_visitor::set_mCenterOfMass)
	    .add_property("inertia", &Body_visitor::get_mInertia, &Body_visitor::set_mInertia)
	    .add_property("spatialInertia", &Body_visitor::get_mSpatialInertia, &Body_visitor::set_mSpatialInertia)
	    .add_property("isVirtual", &Body_visitor::get_mIsVirtual, &Body_visitor::set_mIsVirtual)

	    .def(init<const Body&>(args("Body"), "Copy Constructor"))	// exposing the copy constructor
	    .def(init<double,Math::Vector3d,Math::Matrix3d>(args("mass","center of mass","inertia"), "Constructor from physical parameters like mass, com, interia")) 	// exposing the constructor for creating a body from mass, center of mass, and a 3x3 inertia matrix.  
	    .def(init<const double&, const Math::Vector3d&, const Math::Vector3d&>(args("mass","center of mass","radius of gyration"), "Constructor from physical parameters  like mass, com, radii of gyration")) 	// exposing the constructor for creating a body from mass, center of mass and radii of gyration. 
	    .def("Join",&Body::Join,args("spatial_transform","second_body"),"Join a second body <second_body> at the placement specified by <spatial_transform>")
	    //.def("reassign", &Body::operator=)
        ;
    }
	// Getter for mass
    static double get_mMass(Body& self) { return self.mMass; }
	// Setter for mass
    static void set_mMass(Body& self, const double& mass) { self.mMass = mass; }

	// Getter for COM
    static RigidBodyDynamics::Math::Vector3d get_mCenterOfMass(Body& self) { return self.mCenterOfMass; }
	//Setter for COM
    static void set_mCenterOfMass(Body& self, const Math::Vector3d& com) { self.mCenterOfMass = com; }
	// Getter for inertia
    static RigidBodyDynamics::Math::Matrix3d get_mInertia(Body& self) { return self.mInertia; }
	// Setter for inertia
    static void set_mInertia(Body& self, const Math::Matrix3d& inertia) { self.mInertia = inertia; }

	// Getter for Spatial Inertia
    static RigidBodyDynamics::Math::SpatialMatrix get_mSpatialInertia(Body& self) { return self.mSpatialInertia; }
	// Setter for Spatial Inertia
    static void set_mSpatialInertia(Body& self, const Math::SpatialMatrix spatial_inertia) { self.mSpatialInertia = spatial_inertia; }

	// Getter for mIsVirtual
    static  bool get_mIsVirtual(Body& self){ return self.mIsVirtual; }
	// Setter for mIsVirtual
    static void set_mIsVirtual(Body& self, bool isvirtual){ self.mIsVirtual = isvirtual; }
};

BOOST_PYTHON_MODULE(libbody)
{
  eigenpy::enableEigenPy();

  eigenpy::enableEigenPySpecific<RigidBodyDynamics::Math::Vector3d, Eigen::Vector3d>();
  eigenpy::enableEigenPySpecific<RigidBodyDynamics::Math::Matrix3d, Eigen::Matrix3d>();
  eigenpy::enableEigenPySpecific<RigidBodyDynamics::Math::SpatialMatrix, Eigen::Matrix<double, 6, 6> >();

  class_<RigidBodyDynamics::Math::SpatialTransform>("SpatialTransform", init<>())  
	.def(init<Math::Matrix3d, Math::Vector3d>());	// exposing the constructor for storing a spatial transform built by rotation matrix and translation vector
  //docstring_options doc_options(true);
  class_<Body>("Body", init<>())
        .def(Body_visitor())		
	;
}

File: unittest/body.py
import sys
sys.path.append('/local/skumar/software/rbdlpy/build')
sys.path.append('/local/skumar/devel/lib')

import numpy as np
import libbody as bd
#import libeigenpy

verbose = False

# Exposing the default constructor of Body class
x = bd.Body()
print x.mass
print x.centerOfMass
print x.inertia
print x.spatialInertia
print x.isVirtual

# Exposing the copy constructor of Body class
y = bd.Body(x)	# passing the python object of Body class exposed from c++
print y.mass
print y.centerOfMass
print y.inertia
print y.spatialInertia
print y.isVirtual

# Exposing the constructor for creating a body from mass, center of mass, and a 3x3 inertia matrix.
mass = 5.0
com = np.array([5., 8., 9.])	# make sure elements of array are double
inertia = np.matrix([[1., 2., 3.],[1., 2., 3.],[1., 2., 3.]])	# make sure elements of matrix are double
z = bd.Body(mass, com, inertia)
print z.mass
print z.centerOfMass
print z.inertia
print z.spatialInertia
print z.isVirtual

# Exposing the constructor for creating a body from mass, center of mass, and radii of gyration.
mass = 10.0
com = np.array([2., 8., 5.])
radiusofgyration = np.array([1., 1., 1.])
b = bd.Body(mass, com, radiusofgyration)
print b.mass
print b.centerOfMass
print b.inertia
print b.spatialInertia
print b.isVirtual

rot = np.matrix([[1., 0., 0.],[0., 1., 0.],[0., 0., 1.]])
trans = np.array([1., 1., 1.])
# Exposing the constructor of SpatialTransform class 
st = bd.SpatialTransform(rot, trans)
# Exposing the member function Join which Joins inertial parameters of two bodies to create a composite body. 
b.Join(st, x)
print b.mass

# use of Getter and setter functions
b.mass = 56
print 'mass is manually set to', b.mass 

# Fixed Body class
fb = bd.FixedBody()
assert(fb.mass==0.0)
print fb.centerOfMass
print fb.spatialInertia
assert(fb.MovableParent==0)
print fb.ParentTransform
print fb.BaseTransform
p = fb.CreateFromBody(bd.Body())	# Problem here
